Show icons corresponding to MIME types in the file browser plugin

techee opened the pull request:
The patches are more or less self-explanatory and trivial so there's not much to comment regarding the code itself.

Making ui_get_mime_icon() public also helps my ProjectOrganizer plugin where there's basically a copy of this function.

Of course the "linkage cleanup" patch has a higher priority so if it gets applied soon(ish), I'll update the API part on top of that.

codebrainz commented:
-1 to exposing more internal utility functions that have nothing to do with plugins hooking into Geany. Not like it would be the first, but IMO we should really stop trying to provide a general purpose utility library (glib supplement) as part of the stable/maintained plugin API. 

IMO, we should either try and get such stuff merged upstream or start a new general purpose utility library like GLib that provides all the stuff that's missing, rather than making a fat, all-encompassing general programming library, which increases cost of maintenance, makes the plugin API less focused, and makes it more difficult in the future to change stuff.

(That being said, perhaps that should be a discussion to be had on the dev mailing list?)

techee commented:
I don't actually care about having this particular function part of the API - I can easily duplicate the code (probably with some LOC savings because not all is needed in the filebrowser plugin). What I'd like to see is the feature itself because it's easier to get oriented in the files if you see different icons.

So should I add code similar to ui_get_mime_icon() to the filebrowser plugin?

Regarding the API, yes, it probably makes sense. I saw that similar code would be used three-times so I thought some reuse would be good. But it's true it is a bit too specific and has nothing to do with Geany so yeah, it maybe doesn't belong to the API.

I think it's unrealistic to get this to glib (it wouldn't be useful wide-enough audience) and I don't think it's a good idea to create additional library either (I'm definitely not going to maintain that, it would mean additional dependency for Geany and any library should have some concept and not just offer "possibly useful stuff").

codebrainz commented:
I hate to see redundant code, but it seems so crazy to add stuff like this to the Plugin API. I will raise it on the devel mailing list when I'm done vacation.

I don't want to "block" this PR, I didn't really review it, but it sounds fine in principle, I'm more concerned about the general trend of exposing all kinds of unrelated utility stuff through the plugin API, it's already way too big and un-cohesive, IMO.

techee commented:
I have rebased the patches on top of the current Geany master (after the linkage cleanup merge) and removed the API part so the (slightly reduced) icon code is also present in the plugin but the code duplication isn't that bad I think.

elextr commented:
The Only comment I have is that since g_content_guess_type() sometimes gets a different answer to the filetype Geany opens files as there may be some comments, though most users won't notice.

techee commented:
Yes, one notable difference is *.c and *.h which different file types for g_content_guess_type() while for Geany they are both C types and C icon is shown for both of them. Seeing a different icon for these two is actually one of the main reasons I made this change. Maybe for the Documents tab we could also use something similar and only fallback to Geany filetypes if filetype isn't recognized.

elextr commented:
Have to totally disagree, the *document* tab should show the same filetype as Geany is using for the *document*, note the documents tab filetype changes if you use the document menu to change the filetype of the document.

The filebrowser on the other hand is relating to files that are not open, and in that case g_content_have_a_guess() is as good as anything else.

Since Geany uses file contents as well as extension to determine filetype for the document on opening, I was just noting that its possible that there will be reports that it disagrees with the filebrowser.  I don't think that should stop your improvement to the filebrowser though.

techee commented:
@b4n This is another possible candidate for 1.25. I tested this patch also on Windows/OSX and it seems to work fine there. There's the minor duplication of the code as discussed with Matthew but I think it's not dramatic.

b4n commented on plugins/filebrowser.c:
```diff
@@ -226,10 +261,13 @@ static void add_item(const gchar *name)
+	icon = dir ? g_themed_icon_new("folder") : get_icon(utf8_name);
 	gtk_list_store_set(file_store, &iter,
-		FILEVIEW_COLUMN_ICON, (dir) ? GTK_STOCK_DIRECTORY : GTK_STOCK_FILE,
+		FILEVIEW_COLUMN_ICON, icon,
```
icon has to be unref'd as the tree store will ref it

b4n commented on plugins/filebrowser.c:
```diff
@@ -253,9 +291,10 @@ static void add_top_level_entry(void)
 
 	gtk_list_store_set(file_store, &iter,
-		FILEVIEW_COLUMN_ICON, GTK_STOCK_DIRECTORY,
+		FILEVIEW_COLUMN_ICON, g_themed_icon_new("folder"),
```
ditto, needs unreffing

b4n commented:
LGTM otherwise.

Regarding the Documents tab, I agree with @elextr it should match whatever Geany thinks that file is.  One way already suggested to have different icons for headers is to add a custom filetype for those, setting the `mime_type` as wanted.

elextr commented:
@b4n, yes, headers should be different filetypes for many reasons, their semantics are different, the default commands should be different, using the default .c/.cpp commands compiling overwrites your .o file with a pre-compiled header, which breaks make.  If you build a header it creates the real .pch and that means any edits you make to the .h won't be seen by the compile of the .c/.cpp file.  Hours of fun and confusion results.

</rant>

@techee sorry to "borrow" your pr :)

techee commented:
@b4n Thanks, fixed now.

@elextr I tend to agree they should be a different filetypes (unless there's some catch).

The only problem I can think of is several filetypes sharing the same header extension - this seems to be C, C++, CUDA and Objective-C. I haven't checked the code but is the listing of "*.h" in filetype extensions for each of these files used for something? I guess when opening a file with *.h extension the list is walked from top to bottom and the first language matching the pattern is used (which is C fortunately) so I guess the *.h is useless for the rest anyway.

elextr commented:
@techee yes the first occurrence of *.h in the list will be used, so people who want to use .h as C++ have to be told to delete it from C.
